httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/ChunkAwareResponseEntityProxy.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.client5.http.impl.classic;
+
+import org.apache.hc.client5.http.classic.ExecRuntime;
+import org.apache.hc.core5.function.Supplier;
+import org.apache.hc.core5.http.ClassicHttpResponse;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HttpEntity;
+import org.apache.hc.core5.http.impl.io.ChunkedInputStream;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+class ChunkAwareResponseEntityProxy extends ResponseEntityProxy {
+
+    private ChunkedInputStream internalChunkedInputStream;
+
+    public static void enhanceChunkedResponseSupport(final ClassicHttpResponse response, final ExecRuntime execRuntime) throws IOException {
+        final HttpEntity entity = response.getEntity();
+        if (!(entity.getContent() instanceof ChunkedInputStream)) {
+            throw new IllegalStateException("Underlying stream is not ChunkedInputStream, unable to reach footers.");
+        }
+        final ChunkedInputStream content = (ChunkedInputStream) entity.getContent();
+        if (entity != null && entity.isStreaming() && execRuntime != null) {
+            response.setEntity(new ChunkAwareResponseEntityProxy(entity, execRuntime, content));
+        }
+    }
+
+    ChunkAwareResponseEntityProxy(final HttpEntity entity, final ExecRuntime execRuntime, final ChunkedInputStream chunkedInputStream) {
+        super(entity, execRuntime);
+        this.internalChunkedInputStream = chunkedInputStream;
+    }
+
+    @Override
+    public Supplier<List<? extends Header>> getTrailers() {
```

Comment:
@serkanturgut Could not the `getTrailer` method be added to the `ResponseEntityProxy` eliminating the need for `ChunkAwareResponseEntityProxy`?

Reply:
getTrailers() return a supplier which reaches to ChunkedInputStream.getFooters(). So getTrailers() in ChunkAwareResponseEntityProxy is aware of the underlying stream is ChunkedInputStream. 

If we embed getTrailers() into ResponseEntityProxy, we have to check whether the underlying stream is ChunkedInputStream (ChunkAwareResponseEntityProxy is doing this check during initialization) if it is not a IllegalStateException should be thrown there to notify caller. However this approach contradicts with throw-early-catch-late practice. Given that getTrailers() (or Supplier.get()) typically will be called at the end of HTTP communication by the user of client, It would be a unpleasant surprise to caller to see an exception 

Another reason is that I wanted to leave ResponseEntityProxy stream agnostic as it is currently. Placing new getTrailers() implementation in ResponseEntityProxy brings certain responsibility to ResponseEntityProxy to recognize the type of underlying stream to some extent. 

If you think this is still worth it, I could give it a try to eliminate it. I am open to suggestions, please let me know what you think.

Thank you

Reply:
@serkanturgut I personally see no problem if `ResponseEntityProxy` returned `null` from `getTrailers` if for whatever reason the underlying content stream is not `ChunkedInputStream`. This approach would simplify things considerably. I see no great value in throwing an unchecked exception in the middle of an HTTP message exchange.

Reply:
I added ``getTrailers()`` method into ``ResponseEntityProxy``, removed the if statement in ``MainClientExec`` and did not need another static initializer for ``ResponseEntityProxy``.

The following will be the only addition to the code which will be in ``ResponseEntityProxy``, all other changes in this commit is eliminated now;

```
   @Override
    public Supplier<List<? extends Header>> getTrailers() {
            return new Supplier<List<? extends Header>>() {
                @Override
                public List<? extends Header> get() {
                    final Header[] footers;
                    InputStream underlyingStream = null;
                    try {
                        underlyingStream = getContent();
                    } catch (IOException e) {
                        throw new IllegalStateException("Unable to retrieve input stream");
                    }
                    if (underlyingStream instanceof ChunkedInputStream) {
                        ChunkedInputStream chunkedInputStream = (ChunkedInputStream) underlyingStream;
                        footers = chunkedInputStream.getFooters();
                    } else {
                        footers = new Header[0];
                    }
                    return Arrays.asList(footers);
                }
            };
    }
```

I can't say it looks cleaner however it reduces the amount of changes and need for new class considerably as you mentioned. If this looks good, I will change the unit tests accordingly and push the commit.

Reply:
@serkanturgut As far as I can say simpler is _better_. Looks good to me.